File: 2015/src/day03/main.rs
use std::collections::HashSet;

#[derive(Eq, PartialEq, Hash, Clone)]
struct Point {
    x: i32,
    y: i32,
}

impl Point {
    fn new(x: i32, y: i32) -> Self {
        Self { x, y }
    }
}

fn part1(input: &str) -> usize {
    let mut santa = Point::new(0, 0);
    let mut houses = HashSet::new();
    houses.insert(santa.clone());

    for c in input.chars() {
        if c == '^' {
            santa.y += 1;
        } else if c == 'v' {
            santa.y -= 1;
        } else if c == '<' {
            santa.x -= 1;
        } else if c == '>' {
            santa.x += 1;
        }

        houses.insert(santa.clone());
    }

    houses.len()
}

fn part2(input: &str) -> usize {
    let mut santa = Point::new(0, 0);
    let mut robo_santa = Point::new(0, 0);
    let mut houses = HashSet::new();
    houses.insert(santa.clone());

    for (i, c) in input.chars().enumerate() {
        if c == '^' {
            if i % 2 == 1 {
                santa.y += 1;
            } else {
                robo_santa.y += 1;
            }
        } else if c == 'v' {
            if i % 2 == 1 {
                santa.y -= 1;
            } else {
                robo_santa.y -= 1;
            }
        } else if c == '<' {
            if i % 2 == 1 {
                santa.x -= 1;
            } else {
                robo_santa.x -= 1;
            }
        } else if c == '>' {
            if i % 2 == 1 {
                santa.x += 1;
            } else {
                robo_santa.x += 1;
            }
        }

        if i % 2 == 1 {
            houses.insert(robo_santa.clone());
        } else {
            houses.insert(santa.clone());
        }
    }

    houses.len()
}

fn main() {
    let input = include_str!("input.txt");

    let part1_result = part1(input);
    println!("part1: {}", part1_result);

    let part2_result = part2(input);
    println!("part2: {}", part2_result);
}

File: 2015/day03/main.go
package main

import (
	"fmt"

	"github.com/leakedmemory/aoc/helpers"
)

type point struct {
	x, y int
}

func part1(input string) int {
	set := make(map[point]struct{})
	currPos := point{0, 0}
	set[currPos] = struct{}{}
	for _, c := range input {
		switch c {
		case '>':
			currPos.x++
		case '<':
			currPos.x--
		case '^':
			currPos.y--
		case 'v':
			currPos.y++
		}

		set[currPos] = struct{}{}
	}

	return len(set)
}

func part2(input string) int {
	set := make(map[point]struct{})
	santa := point{0, 0}
	robo := point{0, 0}
	set[santa] = struct{}{}
	for i, c := range input {
		var mover *point
		if i%2 == 0 {
			mover = &santa
		} else {
			mover = &robo
		}

		switch c {
		case '>':
			mover.x++
		case '<':
			mover.x--
		case '^':
			mover.y--
		case 'v':
			mover.y++
		}

		set[*mover] = struct{}{}
	}

	return len(set)
}

func main() {
	input := helpers.ReadFile("2015/day03/input.txt")
	fmt.Printf("part 1: %d\n", part1(input))
	fmt.Printf("part 2: %d\n", part2(input))
}
